### PR TITLE
Maml++

### LINKED ISSUE #261: MAML++
Hi 
Thank you for the great library and examples. it would be awesome if you could integrate MAML++ (ICLR 2019).

@inproceedings{
antoniou2018how,
title={How to train your {MAML}},
author={Antreas Antoniou and Harrison Edwards and Amos Storkey},
booktitle={International Conference on Learning Representations},
year={2019},
url={https://openreview.net/forum?id=HJGven05Y7},
}
image

### PR DESCRIPTION
**Pull request by DubiousCactus**
### Description

Fixes #261 

Work in Progress implementation of [MAML++](https://arxiv.org/abs/1810.09502).

### TODO:
- [x] Multi-Step Loss for improved gradient stability
- [x] Derivative-Order Annealing for reduced second order derivative cost
- [x] Per-Step Batch Normalization Running Statistics
- [x] Per-Step Batch Normalization Weights and Biases
- [x] Learning Per-Layer Per-Step Learning Rates and Gradient Directions
- [x] Cosine Annealing of Meta-Optimizer Learning Rate

### Contribution Checklist

If your contribution modifies code in the core library (not docs, tests, or examples), please fill the following checklist.

- [x] My contribution is listed in CHANGELOG.md with attribution.
- [ ] My contribution modifies code in the main library.
- [ ] My modifications are tested.
- [x] My modifications are documented.

### CODE REVIEW

**Comment by kzhang2** [comment]
Any updates on this?

**Comment by DubiousCactus** [comment]
> Any updates on this?

Yeah sorry, I've been very busy lately... I started looking into "Per-Step Batch Normalization Running Statistics", but that's not easy to integrate into this library because it has to be implemented as a custom batch norm layer module. I'll integrate it this evening :)

**Comment by DubiousCactus** [comment]
So I've implemented BRNS and BNWB, although it still needs to be properly tested. I had to create a new BatchNorm layer, as in the original implementation, and therefore a new CNN class that uses it. I'm not sure what's the best way to integrate this batch norm class, and if it should be more flexible (and if so, how to parameterise it), so I'd gladly take your criticism.
I will try and implement the last improvement, LSLR, maybe tomorrow.

**Comment by DubiousCactus** [comment]
I'll fix the tests failing today or tomorrow. Will probably just create a MAMLpp wrapper to avoid adding parameters to MAML.

File: learn2learn/algorithms/maml.py [hunk 1]
```diff
@@ -103,14 +107,32 @@ def __init__(self,
...
+                        if param.requires_grad
+                    ]
+                )
+                > 0
+            ):
+                lslr[layer_name.replace(".", "-")] = torch.nn.Parameter(
+                    data=torch.ones(adaptation_steps) * init_lr,
+                    requires_grad=True,
+                )
+        return lslr
```
Comment (seba-1511) [inline learn2learn/algorithms/maml.py:130]: @DubiousCactus this should be removed, right? In general, can you stash this file so it keeps the original formatting?

File: CHANGELOG.md [hunk 2]
```diff
@@ -20,8 +20,13 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 * Bounding box cropping for Aircraft and CUB200.
 * Pretrained weights for vision models with: `l2l.vision.models.get_pretrained_backbone()`.
+* New vision example: MAML++. (@[DubiousCactus](https://github.com/DubiousCactus))
 * Add `keep_requires_grad` flag to `detach_module`. ([Zhaofeng Wu](https://github.com/ZhaofengWu))
 
+### Changed
+
+* Implement LSLR (MAML++) in the MAML module. (@[DubiousCactus](https://github.com/DubiousCactus))
+
```
Comment (seba-1511) [inline CHANGELOG.md:29]: Move to Unreleased.

**Review by seba-1511 — COMMENTED** [review]
@DubiousCactus This looks great, thanks a lot for the changes.

I've added 2 comments:

1. revert maml.py to its original content, and
2. formatting and moving the changelog contributions to unreleased.

Once those are done the PR is ready to be merged.

File: learn2learn/algorithms/maml.py (hunk 1 above)
Comment (DubiousCactus) [reply learn2learn/algorithms/maml.py]: Yes, sorry about that.

File: CHANGELOG.md (hunk 2 above)
Comment (DubiousCactus) [reply CHANGELOG.md]: I removed the LSLR implementation since it's not relevant anymore.

**Comment by seba-1511** [comment]
Thanks a lot for the changes @DubiousCactus. Looks good to merge.
